src/LandIce/evaluators/LandIce_HydrologyMeltingRate_Def.hpp

```diff
@@ -71,43 +73,64 @@ HydrologyMeltingRate (const Teuchos::ParameterList& p,
   } else {
     layout = dl->qp_scalar;
   }
-  u_b  = PHX::MDField<const IceScalarT>(p.get<std::string> ("Sliding Velocity Variable Name"), layout);
-  G    = PHX::MDField<const ParamScalarT>(p.get<std::string> ("Geothermal Heat Source Variable Name"), layout);
-  beta = PHX::MDField<const ScalarT>(p.get<std::string> ("Basal Friction Coefficient Variable Name"), layout);
-  m    = PHX::MDField<ScalarT>(p.get<std::string> ("Melting Rate Variable Name"),layout);
-
-  this->addDependentField(beta);
-  this->addDependentField(u_b);
-  this->addDependentField(G);
+
+  m = PHX::MDField<ScalarT>(p.get<std::string> ("Melting Rate Variable Name"),layout);
   this->addEvaluatedField(m);
 
-  this->setName("HydrologyMeltingRate"+PHX::print<EvalT>());
+  Teuchos::ParameterList& hy_pl = *p.get<Teuchos::ParameterList*>("LandIce Hydrology");
+  Teuchos::ParameterList& melt_pl = hy_pl.sublist("Melting Rate");
+
+  if (melt_pl.isParameter("Use Given Value")) {
+    m_value = melt_pl.get("Use Given Value",0.0);
+    m_given = true;
+  } else {
+    if (melt_pl.get("Use Friction Melt",true)) {
+      u_b  = PHX::MDField<const IceScalarT>(p.get<std::string> ("Sliding Velocity Variable Name"), layout);
+      beta = PHX::MDField<const ScalarT>(p.get<std::string> ("Basal Friction Coefficient Variable Name"), layout);
+      this->addDependentField(beta);
+      this->addDependentField(u_b);
+
+      friction = true;
+    }
+    if (melt_pl.get("Use Geothermal Melt", true)) {
+      G = PHX::MDField<const ParamScalarT>(p.get<std::string> ("Geothermal Heat Source Variable Name"), layout);
+      this->addDependentField(G);
+
+      G_field = true;
+    } else if (melt_pl.isParameter("Given Geothermal Flux")) {
+      G_given = true;
+      G_value = melt_pl.get<double>("Given Geoethermal Flux");
+    }
+  }
+
+  this->setName("HydrologyMeltingRate" + (nodal ? std::string("Nodal") : std::string("QPs")) +PHX::print<EvalT>());
 }
 
 //**********************************************************************
 template<typename EvalT, typename Traits, bool IsStokes>
 void HydrologyMeltingRate<EvalT, Traits, IsStokes>::
-postRegistrationSetup(typename Traits::SetupData d,
-                      PHX::FieldManager<Traits>& fm)
+postRegistrationSetup(typename Traits::SetupData /* d */,
+                      PHX::FieldManager<Traits>& )
 {
-  this->utils.setFieldData(u_b,fm);
-  this->utils.setFieldData(G,fm);
-  this->utils.setFieldData(beta,fm);
-  this->utils.setFieldData(m,fm);
+  m.deep_copy(m_value);
 }
 
 //**********************************************************************
 template<typename EvalT, typename Traits, bool IsStokes>
 void HydrologyMeltingRate<EvalT, Traits, IsStokes>::evaluateFields (typename Traits::EvalData workset)
 {
-  // m = \frac{ G - \beta |u_b|^2 + \nabla (phiH-N)\cdot q }{L} %% The nonlinear term \nabla (phiH-N)\cdot q can be ignored
+  // m = ( G - \beta |u_b|^2 ) / L
```

Comment:
In the hydrology equation we have `m/rho_w` and `m/rho_i` in the mass and cavity equations respectively. Though melting in the cavity equation is usually off, I wanted to support it in ALI (models are quite...fluid, and keep changing from paper to paper, but I try to allow all options if possible). That's why I compute the melting without dividing by rho, so that I can use m in both equations, if needed.

I can add the units.

Reply:
I see. It makes sense to me. I might reconsider what we do for the temperature.. in fact we have terms like rho_i/rho_w for the water in the temperate ice